drivers/sensor/lsm6dsr/lsm6dsr.c

```diff
@@ -0,0 +1,803 @@
+	}
+
+	if (lsm6dsr_gyro_set_odr_raw(dev, odr) < 0) {
+		LOG_DBG("failed to set gyroscope sampling rate");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int lsm6dsr_gyro_range_set(const struct device *dev, int32_t range)
+{
+	lsm6dsr_fs_g_t fs;
+
+	switch (range) {
+	case 0: {
+		fs = LSM6DSR_250dps;
+	} break;
+	case 1: {
+		fs = LSM6DSR_4000dps;
+	} break;
+	case 2: {
+		fs = LSM6DSR_125dps;
+	} break;
+	case 4: {
+		fs = LSM6DSR_500dps;
+	} break;
+	case 8: {
+		fs = LSM6DSR_1000dps;
+	} break;
+	case 12: {
+		fs = LSM6DSR_2000dps;
+	} break;
+	default:
+		return -EINVAL;
+	}
+
+	if (lsm6dsr_gyro_set_fs_raw(dev, fs) < 0) {
+		LOG_DBG("failed to set gyroscope full-scale");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int lsm6dsr_gyro_pm_set(const struct device *dev, uint8_t gyro_pm)
+{
+	const struct lsm6dsr_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	lsm6dsr_g_hm_mode_t mode;
+
+	switch (gyro_pm) {
+	case 0: { /* High Performance mode */
+		mode = LSM6DSR_GY_HIGH_PERFORMANCE;
+	} break;
+	case 1: { /* Low/Normal Power mode */
+		mode = LSM6DSR_GY_NORMAL;
+	} break;
+	default:
+		return -EINVAL;
+	}
+	if (lsm6dsr_gy_power_mode_set(ctx, mode) < 0) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int lsm6dsr_gyro_config(const struct device *dev,
+			       enum sensor_channel chan,
+			       enum sensor_attribute attr,
+			       const struct sensor_value *val)
+{
+	switch (attr) {
+	case SENSOR_ATTR_FULL_SCALE:
+		return lsm6dsr_gyro_range_set(dev, sensor_rad_to_degrees(val));
+	case SENSOR_ATTR_SAMPLING_FREQUENCY:
+		return lsm6dsr_gyro_odr_set(dev, val);
+	case SENSOR_ATTR_CONFIGURATION:
+		return lsm6dsr_gyro_pm_set(dev, val->val1);
+	default:
+		LOG_DBG("Gyro attribute not supported.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int lsm6dsr_attr_set(const struct device *dev,
+			    enum sensor_channel chan,
+			    enum sensor_attribute attr,
+			    const struct sensor_value *val)
+{
+	switch (chan) {
+	case SENSOR_CHAN_ACCEL_XYZ: {
+		return lsm6dsr_accel_config(dev, chan, attr, val);
+	}
+	case SENSOR_CHAN_GYRO_XYZ: {
+		return lsm6dsr_gyro_config(dev, chan, attr, val);
+	}
+	default:
+		LOG_WRN("attr_set() not supported on this channel.");
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int lsm6dsr_sample_fetch_accel(const struct device *dev)
+{
+	const struct lsm6dsr_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm6dsr_data *data = dev->data;
+
+	if (lsm6dsr_acceleration_raw_get(ctx, data->acc) < 0) {
+		LOG_DBG("Failed to read sample");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int lsm6dsr_sample_fetch_gyro(const struct device *dev)
+{
+	const struct lsm6dsr_config *cfg = dev->config;
+	stmdev_ctx_t *ctx = (stmdev_ctx_t *)&cfg->ctx;
+	struct lsm6dsr_data *data = dev->data;
+
+	if (lsm6dsr_angular_rate_raw_get(ctx, data->gyro) < 0) {
+		LOG_DBG("Failed to read sample");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#if defined(CONFIG_LSM6DSR_ENABLE_TEMP)
```

Comment:
If you're using `if (IS_ENABLED(CONFIG_LSM6DSR_ENABLE_TEMP)) { ... ` elsewhere, then the `#if ... #endif` is not needed here, which makes code a bit easier to read (and it is removed by the linker).

Reply:
doing this change make `struct lsm6dsr_data` to also define `int16_t temp_sample;`, otherwise there will be warnings. I was thinking that using: `IF_ENABLED(CONFIG_FLAG, int16_t temp_sample;)` would fix that but it's not working like that.